core/mm/core_mmu.c

```diff
@@ -1003,18 +1024,29 @@ static bool assign_mem_va(vaddr_t tee_ram_va,
 					       CORE_MMU_PGDIR_MASK;
 
 				if (ADD_OVERFLOW(va, offs, &va))
-					return false;
+					goto maybe_retry;
 			}
 
 			map->va = va;
 			if (ADD_OVERFLOW(va, map->size, &va))
-				return false;
+				goto maybe_retry;
 			if (va >= BIT64(core_mmu_get_va_width()))
-				return false;
+				goto maybe_retry;
 		}
 	}
 
 	return true;
+
+maybe_retry:
+	if (IS_ENABLED(CFG_WITH_PAGER) && !retry &&
+	    tee_ram_at_top != place_tee_ram_at_top(tee_ram_va)) {
+		/* Retry to setup mapping consumming an extra xlat table */
+		tee_ram_at_top = place_tee_ram_at_top(tee_ram_va);
+		retry = true;
+		goto define_vmap;
```

Comment:
This is not how loops are implemented. Please find another way.

Reply:
i agree it is not clean. i'll update...

Reply:
I've split `assign_mem_va()` in 2 functions to ease integration.